clang/lib/CodeGen/CGDecl.cpp

```diff
@@ -1353,6 +1353,19 @@ void CodeGenFunction::EmitLifetimeEnd(llvm::Value *Size, llvm::Value *Addr) {
   C->setDoesNotThrow();
 }
 
+void CodeGenFunction::EmitFakeUse(Address Addr) {
+  // We do not emit a fake use if we want to apply optnone to this function,
+  // even if we might not apply it anyway due to minsize or similar attributes.
+  if (!CGM.getCodeGenOpts().DisableO0ImplyOptNone &&
+      CGM.getCodeGenOpts().OptimizationLevel == 0)
+    return;
```

Comment:
Yes - that's how it originally worked, and the choice being moved to here is an accidental hangover from some earlier attempts at modifications I made; specifically I tried to make it so that fake uses would only be applied if the specific function did not have `optnone`, but the decision-making that applies `optnone` to functions happens after we finalize fake uses. I'm currently considering whether it's a good idea to just enable `-fextend-lifetimes` at O0, in which case this behaviour can be removed outright, otherwise it could be moved back to the driver as you've suggested.

Reply:
There's no value to enabling `-fextend-lifetimes` at O0, because we only emit stack-home locations at O0, which are valid for the entire function. However, I don't remember whether that decision is a blanket decision based on opt level or per-function based on optnone. If it's based on opt level, then adding fake-uses to an optnone function (at O1 and up) might actually be the right thing to do.

Reply:
(FWIW - clang IRGen always emit stack-home locations, at any optimization level - it's only a question of whether we run the optimizations (& whether they fire, eg: we might run the optimization pass, but it detects a function is optnone, then it does nothing) that remove those stack homes)

Reply:
Objection withdrawn. Because it's opt-in, we should do what the user requested. As you say, in a normal `-O0` situation all this does is increase IR size for no benefit, but it has its uses.

Reply:
(might be worth removing "-disable-O0-optnone" as it's a bit of a hazard for reasons like this - if folks want optimizable IR, but without optimizations applied, they'd probably be better off disabling optimizations (`-disable-llvm-optzns`) I think?)